### DISCUSSION TITLE
[Bug] Unable to install on HPC. GRASS 8.0

### DISCUSSION THREAD

**Discussion by ZehaoL45** [OSGeo/grass#2115] [Q&A]
**Describe the bug**
I've been trying to install GRASS on my HPC account but it always shows the same message.

This is the HPC I am using. https://www.rc.ucl.ac.uk/docs/Clusters/Kathleen/

**To Reproduce**

Below is the code I ran:
# compile script

module purge
module load beta-modules
module load r/r-4.1.1_bc-3.13
module load nano/2.4.2
module load bison/3.0.4/gnu-4.9.2
module list

#Define some variables, you need to add these so that the paths are defined cor$
#the programs and libraries are installed in the place you want them to be.
#The way the script works is to make a master directory then add the path to th$
#then run the install in there and it *should*
#make 2 folders where your bins and libs should be for each of the things
#compiled things in this script

MASTER_FOLDER=/home/uczl875/grass_latest
ZSTD_LIBRARY=$MASTER_FOLDER/zstd_install

INSTALL_PREFIX=$MASTER_FOLDER/grass_install
mkdir $MASTER_FOLDER/grass_install
mkdir $MASTER_FOLDER/zstd_install

#Installing the zstd compression library

git clone https://github.com/facebook/zstd.git
cd zstd/
make ; make install prefix=$MASTER_FOLDER/zstd_install
export PATH=$ZSTD_LIBRARY/bin:$PATH
export LD_LIBRARY_PATH=$ZSTD_LIBRARY/lib:$LD_LIBRARY_PATH
cd $MASTER_FOLDER

#Installing GRASS 80
git clone https://github.com/OSGeo/grass.git grass_master
cd $MASTER_FOLDER/grass_master

export PATH=/home/uczl875/grass_latest/zstd/bin:$PATH
export LD_LIBRARY_PATH=/home/uczl875/grass_latest/zstd_install/bin:$LD_LIBRARY_PATH

export PATH=/home/uczl875/grass_latest/zstd/lib:$PATH
export LD_LIBRARY_PATH=/home/uczl875/grass_latest/zstd_install/lib:$LD_LIBRARY_PATH

pwd
./configure --prefix=$INSTALL_PREFIX/ --with-zstd-includes=$ZSTD_LIBRARY/include --with-zstd-libs=$ZSTD_LIBRARY/lib

make ; make install

**Error messages:**
 Started compilation: 2021年 11月 11日 星期四 18:18:40 GMT
Errors in:
/lustre/home/uczl875/grass_latest/grass_master/db/drivers/dbf
/lustre/home/uczl875/grass_latest/grass_master/raster

In case of errors please change into the directory with error and run 'make'.
If you get multiple errors, you need to deal with them in the order they
appear in the error log. If you get an error building a library, you will
also get errors from anything which uses the library.

Finished compilation: 2021年 11月 11日 星期四 18:32:42 GMT

**Then I go into the folders but it still shows this:**

[uczl875@login02 grass_master]$ cd db/drivers/dbf
[uczl875@login02 dbf]$ make
: && gcc -L/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -L/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -Wl,--export-dynamic -Wl,-rpath-link,/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -o /lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/driver/db/dbf OBJ.x86_64-pc-linux-gnu/column.o OBJ.x86_64-pc-linux-gnu/create_table.o OBJ.x86_64-pc-linux-gnu/cursor.o OBJ.x86_64-pc-linux-gnu/db.o OBJ.x86_64-pc-linux-gnu/dbfexe.o OBJ.x86_64-pc-linux-gnu/describe.o OBJ.x86_64-pc-linux-gnu/driver.o OBJ.x86_64-pc-linux-gnu/error.o OBJ.x86_64-pc-linux-gnu/execute.o OBJ.x86_64-pc-linux-gnu/fetch.o OBJ.x86_64-pc-linux-gnu/listtab.o OBJ.x86_64-pc-linux-gnu/main.o OBJ.x86_64-pc-linux-gnu/select.o OBJ.x86_64-pc-linux-gnu/str.o OBJ.x86_64-pc-linux-gnu/table.o    -lgrass_dbmidriver.8.0 -lgrass_dbmibase.8.0 -lgrass_gis.8.0 -lgrass_shape.8.0 -lgrass_sqlp.8.0   -lm
/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib/libgrass_sqlp.8.0.so: undefined reference to `yyerror'
/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib/libgrass_sqlp.8.0.so: undefined reference to `yylex'
collect2: error: ld returned 1 exit status
make: *** [/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/driver/db/dbf] Error 1

[uczl875@login02 grass_master]$ cd raster
[uczl875@login02 raster]$ make
make -C r.mapcalc
make[1]: Entering directory `/lustre/home/uczl875/grass_latest/grass_master/raster/r.mapcalc'
: && gcc -L/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -L/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -Wl,--export-dynamic -Wl,-rpath-link,/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/lib -o /lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/bin/r.mapcalc OBJ.x86_64-pc-linux-gnu/column_shift.o OBJ.x86_64-pc-linux-gnu/evaluate.o OBJ.x86_64-pc-linux-gnu/expression.o OBJ.x86_64-pc-linux-gnu/function.o OBJ.x86_64-pc-linux-gnu/main.o OBJ.x86_64-pc-linux-gnu/map.o OBJ.x86_64-pc-linux-gnu/mapcalc.tab.o OBJ.x86_64-pc-linux-gnu/mapcalc.yy.o OBJ.x86_64-pc-linux-gnu/xarea.o OBJ.x86_64-pc-linux-gnu/xcoor.o OBJ.x86_64-pc-linux-gnu/xres.o OBJ.x86_64-pc-linux-gnu/xrowcol.o    -lgrass_calc.8.0 -lgrass_gis.8.0 -lgrass_raster.8.0 -lgrass_btree.8.0        -lm
OBJ.x86_64-pc-linux-gnu/mapcalc.tab.o: In function `yyparse':
/lustre/home/uczl875/grass_latest/grass_master/raster/r.mapcalc/mapcalc.tab.c:1308: undefined reference to `yylex'
OBJ.x86_64-pc-linux-gnu/mapcalc.tab.o: In function `parse_string':
/lustre/home/uczl875/grass_latest/grass_master/raster/r.mapcalc/mapcalc.y:270: undefined reference to `initialize_scanner_string'
OBJ.x86_64-pc-linux-gnu/mapcalc.tab.o: In function `parse_stream':
/lustre/home/uczl875/grass_latest/grass_master/raster/r.mapcalc/mapcalc.y:278: undefined reference to `initialize_scanner_stream'
collect2: error: ld returned 1 exit status
make[1]: *** [/lustre/home/uczl875/grass_latest/grass_master/dist.x86_64-pc-linux-gnu/bin/r.mapcalc] Error 1
make[1]: Leaving directory `/lustre/home/uczl875/grass_latest/grass_master/raster/r.mapcalc'
make: *** [default] Error 2

**Expected behavior**
I am guessing that this is related to the version of GRASS?

**Comment by neteler** [comment]
Please add (as per the issue form) details about your software components:
    - run `g.version -rge` in a GRASS GIS terminal session and post the output here

I suppose that you can start a session nonetheless the reported compilation errors (`db/drivers/dbf` is probably not quite relevant anyway).

**Comment by ZehaoL45** [comment]
I'd like to, but I cannot run the g.version -rge  since I haven't had my GRASS installed yet because of the installation error..

Zehao Li
Research Student
UCL Institute of Archaeology
31-34 Gordon Square
London, WC1H 0PY
0044 7384392051 (UK)
0086 15510275123 (CN)
***@***.***

**Comment by neteler** [comment]
> I'd like to, but I cannot run the g.version -rge since I haven't had my GRASS installed yet because of the installation error.. 

No need to install, you can run it from

`bin.x86_64-pc-linux-gnu/grass`

**Comment by ZehaoL45** [comment]
it is grass78

Also I load the modules below to when doing the installation:

flex 2.5.39
bison 3.0.4/gnu-4.9.2
gdal 3.3.2/gnu-10.2.0
python 3.9
gcc 10.2.0

I think the yyerrors() and yyparse() are related to bison.

**Reply by neteler** [reply]
I suspect that the flex version plays a role here.

Someone reports that downgrading the version helped (in a different software project): https://stackoverflow.com/a/64828130/452464

However, on my Fedora 35 box I use successfully:

```
flex --version
flex 2.6.4

```

**Comment by wenzeslaus** [comment]
Not directly solving your problem, but scripts I use to install on our HPC are here (I use conda for dependencies): https://github.com/ncsu-geoforall-lab/grass-gis-on-hpc-henry2

**Comment by wenzeslaus** [comment]
It is possible to install GRASS GIS on HPC in general and this is too localized to be an reproducible issue. Moving to discussion where we can continue this if needed.

**Comment by ste-goldstein** [comment]
I'm having the exact same issue trying to install it on Arch Linux from the [AUR Grass package](https://aur.archlinux.org/packages/grass).
Is there a fix ?

**Reply by wenzeslaus** [reply]
Speaking in context of the discussion at AUR, it seems the version combination is not really a problem by itself. I have Flex 2.6.4 and I compile 8.2.0 from Git without an issue on Ubuntu, but my flex is simply `/usr/bin/flex` (detected in `config.log` as `found /usr/bin/flex`).

**Reply by ste-goldstein** [reply]
I've been stuck with this for months now, without being able to find a solution.
Any help is really appreciated.
```
$ which flex
/usr/bin/flex
$ flex --version
flex 2.6.4
```

**Reply by marisn** [reply]
I have no problems on Debian with flex 2.6.4
Please provide your config.log file. Also check your env for anything bison/yacc/flex related.

**Comment by nilason** [comment]
Please try:
```bash
cd lib/db/sqlp
make clean && make

cd ../../../raster/r.mapcalc
make clean && make
```
and report the output.

**Reply by cwhite911** [reply]
I had the same issue and this fixed the issue for me.

**Comment by ste-goldstein** [comment]
Thanks for your help. This is really weird, but at my at least 30th attempt to clean build it, it worked.
In my previous trials, I  re-installed flex many times (with and without uninstalling it first) , to no avail.

This time I downgraded and upgraded it back between AUR releases (2.6.4-2 and 2.6.4-3) and it finally worked.
The config .log and commands suggested by @nilason obviously now are fine.

I found however the old config.log from another trial where it failed, and indeed flex wasn't being found 🤔
```
configure:5053: checking for flex
configure:5080: result: no
```
[config.log](https://github.com/OSGeo/grass/files/10039063/config.log)

**Reply by neteler** [reply]
> I found however the old config.log from another trial where it failed, and indeed flex wasn't being found thinking
> 
> ```
> configure:5053: checking for flex
> configure:5080: result: no
> ```

In this case the output of `which flex` would be interesting - is it in `$PATH`?

**Reply by ste-goldstein** [reply]
Right now it is at `/usr/bin/flex`, so it's in `$PATH`. Don't think I can find out what it was before, when the build was failing.
I wish it still didn't work so we could find the issue lol
